projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs

```diff
@@ -0,0 +1,158 @@
+                ActivityContext.TryParse(traceId, traceState, out ActivityContext parentContext);
+                Activity activity = StartRabbitMQActivity($"{routingKey} receive", ActivityKind.Consumer, parentContext);
+                if (activity != null && activity.IsAllDataRequested)
+                {
+                    if (baggage != null)
+                    {
+                        foreach (var item in baggage)
+                        {
+                            Activity.Current?.SetBaggage(item.Key, item.Value);
+                        }
+                    }
+
+                    PopulateMessagingTags("receive", routingKey, exchange, deliveryTag, readOnlyBasicProperties, bodySize, activity);
+                }
+
+                return activity;
+            }
+
+            return null;
+        }
+
+        internal static Activity Process(BasicDeliverEventArgs deliverEventArgs)
+        {
+            if (s_source.HasListeners())
+            {
+                // Extract the PropagationContext of the upstream parent from the message headers.
+                DistributedContextPropagator.Current.ExtractTraceIdAndState(deliverEventArgs.BasicProperties, ExtractTraceIdAndState, out string traceId, out string traceState);
+                IEnumerable<KeyValuePair<string, string>> baggage = DistributedContextPropagator.Current.ExtractBaggage(deliverEventArgs.BasicProperties, ExtractTraceIdAndState);
+                ActivityContext.TryParse(traceId, traceState, out ActivityContext parentContext);
+                Activity activity = StartRabbitMQActivity($"{deliverEventArgs.RoutingKey} process", ActivityKind.Consumer, parentContext);
+                if (activity != null && activity.IsAllDataRequested)
+                {
+                    if (baggage != null)
+                    {
+                        foreach (var item in baggage)
+                        {
+                            Activity.Current?.SetBaggage(item.Key, item.Value);
+                        }
+                    }
+
+                    PopulateMessagingTags("process", deliverEventArgs.RoutingKey, deliverEventArgs.Exchange, deliverEventArgs.DeliveryTag, deliverEventArgs.BasicProperties, deliverEventArgs.Body.Length, activity);
+                }
+
+                return activity;
+            }
+
+            return null;
+        }
+
+        private static Activity StartRabbitMQActivity(string name, ActivityKind kind, ActivityContext parentContext = default)
+        {
+            Activity activity = s_source.CreateActivity(name, kind, parentContext, idFormat: ActivityIdFormat.W3C)?.Start();
+            if (activity != null && activity.IsAllDataRequested)
+            {
+                if (!string.IsNullOrEmpty(Thread.CurrentThread.Name))
+                {
+                    activity.SetTag("thread.name", Thread.CurrentThread.Name);
```

Comment:
I would propose to remove it from rabbit instrumentation and instead rely on users (OTel extension, OTel distro) to add a common activity processor to populate `thread.*` attributes.